### DISCUSSION TITLE
blog/thinking-in-react-query

### DISCUSSION THREAD

**Discussion by giscus** [TkDodo/blog-comments#96] [Announcements]
# blog/thinking-in-react-query

In this talk, we will learn how a different mindset can help us understand React Query and work with it efficiently.

https://tkdodo.eu/blog/thinking-in-react-query

**Comment by tlatmdxo123** [comment]
Thanks to you, I've become able to use React Query properly! I appreciate it. While reading your post, I had one question. If we need to fetch data from the server and edit that data through a form, the state in question seems to be both a server state and a client state. In such cases, how can we manage it with React Query?

**Reply by wporoslo** [reply]
https://tkdodo.eu/blog/react-query-and-forms

**Comment by BennieCopeland** [comment]
How would you handle a HATEOAS REST API? Where the client application only knows the root API endpoint, and then must query and navigate from the hypermedia links found on those responses?

**Reply by TkDodo** [reply]
good question. I think it would be some form of dependent query. let's take this example response from [the wikipedia page on HATEOS](https://en.wikipedia.org/wiki/HATEOAS):

```
{
    "account": {
        "account_number": 12345,
        "balance": {
            "currency": "usd",
            "value": 100.00
        },
        "links": {
            "deposits": "/accounts/12345/deposits",
            "withdrawals": "/accounts/12345/withdrawals",
            "transfers": "/accounts/12345/transfers",
            "close-requests": "/accounts/12345/close-requests"
        }
    }
}
```

we would probably have a `useAccountQuery(accountNumber)`. If we then wanted to get deposits, I would make a `useDeposits` query that would also get the account number passed. In there, I would call `useAccountQuery` and use `links.deposits` to get the deposits. We would then just disable that query for as long as we have no account information:

```
const { data } = useAccountQuery('12345')
const depositsUrl = data?.links.deposits

useQuery({
  queryKey: ['deposits', depositsUrl],
  queryFn: () => axios.get(depositsUrl),
  enabled: !!depositsUrl
})
```

**Reply by BennieCopeland** [reply]
Oh, I hadn't thought of using the url as part of the queryKey. That's nice. Would you compose those into a larger `useDepositsQuery` hook? Or use it like that in a component? Right now, I'm composing them into hooks like so:

```typescript
const getApi = async () => {
    const response = await axios.get(API_URL);
    return apiSchema.parse(response.data);
}

const apiQuery = {
    queryKey: ['api'],
    queryFn: getApi
}

const useApi = () => {
    return useQuery(apiQuery);
}

const getTrainingEvents = async (loggedInApi: LoggedInApi | undefined) => {
    if (typeof api === "undefined")
        return Promise.reject(new Error("Invalid api object"));

    const response = await axios.get(loggedInApi._links["tat:training-events"].href);

    return TrainingEventsSchema.parse(response.data);
}

const trainingEventsQuery = (loggedInApi: LoggedInApi | undefined) => ({
    queryKey: ["training-events"],
    queryFun: () => getTrainingEvents(loggedInApi),
    enabled: !!loggedInApi
})

const useTrainingEvents = () => {
    const { data } = useApi();
    const loggedInApi = api as LoggedInApi; // Felt a union was better than nullable links for the root API

    return useQuery(trainingEventsQuery(loggedInApi));
}

// or

const useTrainingEvents = (loggedInApi: LoggedInApi | undefined) => {
    return useQuery(trainingEventsQuery(loggedInApi));
}

// also trying to mix this in with react router and loader's/action's

const loader = (queryClient: QueryClient) => async () => {
    const api = queryClient.getQueryData<Api>(apiQuery.queryKey) ?? (await queryClient.fetchQuery(apiQuery));
    const loggedInApi = api as LoggedInApi;
    const query = trainingEventsQuery(loggedInApi);

    return (queryClient.getQueryData<TrainingEvents>(query.queryKey) ?? (await queryClient.fetchQuery(query))
}

```

**Comment by danbeo95** [comment]
Great post
Thank for your sharing
I learn a lot thing from this post

**Comment by consdu** [comment]


**Reply by TkDodo** [reply]
No 😅🤷‍♂️

**Reply by consdu** [reply]
@TkDodo nevermind 😆 doesn't seem to let me delete the comment, maybe you can
